content/vault/v1.21.x (rc)/content/docs/updates/important-changes.mdx

```diff
@@ -49,6 +49,38 @@ vault write auth/kubernetes/role/demo \
 Refer to the [Kubernetes authentication docs](/vault/docs/auth/kubernetes) for
 more information.
 
+### Item-by-item list comparison for allowed_parameters and denied_parameters ((#allowed-parameters-list))
+
+| Change       | Affected version | Vault edition
+| ------------ | ---------------- | -------------
+| Breaking     | 1.21.0           | All
+
+Prior to Vault 1.21, when a parameter in a request was subject to
+`allowed_parameters` or `denied_parameters` checks and the parameter's value was
+a list, Vault evaluated the **entire list** as a single value.
+This meant Vault would only find a match if the **whole list** exactly matched
+an entry in `allowed_parameters` or `denied_parameters`, which was often
+unintuitive.
+
+Starting with Vault 1.21, this behavior has changed.
+When a parameter's value is a list, **each element** in that list is now checked
+individually against the allowed or denied values.
```

Comment:
```suggestion
Vault now checks each value in a list parameter against allowed/deined values
in the applicable Vault policy and allows or denies requests if the policy
defines the list as a whole or every individual element of the list. For
example, if the request includes a list like `['a', 'b', 'c']`, it matches to a
policy that includes `['a', 'b', 'c']` or the individual values `a`, `b`, and
`c`.

Previous versions of Vault only matched list parameters when the associated
policy defined the list as a whole. As a result, Vault allowed lists containing
denied values as long as the policy did not deny that exact list and denied
lists containing allowed values because the policy did not allow the exact list.
```
Style correction: write in active voice

Reply:
>and allows or denies requests if the policy
defines the list as a whole or every individual element of the list

that's not quite right, vault will allow if all values in the list match the allowed values, but will deny if **any** values in the list match a denied value. That same caveat also makes it hard to smoothly use the example, not sure how to integrate it nicely into the paragraph...

Reply:
sorry, this version doesn't mention the fields to which the breaking change applies at all, I think that's very important.

also
>allows or denies requests if the policy defines the list as a whole or every individual element of the list

the "or every element of the list" is just for `allowed_parameters`, for `denied_parameters` it's "or any element of the list"